### DISCUSSION TITLE
When 3 tunnels are formed with 7 Intermediate CA's on each site, all 7 Intermediate CA's are not listed in swanctl --list-cert output for few of the tunnels

### DISCUSSION THREAD

**Discussion by demahaj1** [strongswan/strongswan#899] [Q&A] [answered]
In swanctl --list-cert output on Primary site, sometimes one of the remote endpoint certificate and few Intermediate CA's are not getting listed, Mostly this issue was observed during reauthentication, although all tunnels are intact and visible in swanctl --list-sa output

Can you please provide reason for above observed behavior. Attaching logs and certificate list output.. Thanks

[cert.txt](https://github.com/strongswan/strongswan/files/8068362/cert.txt)
[charon.log](https://github.com/strongswan/strongswan/files/8068363/charon.log)

**Comment by tobiasbrunner** [comment, marked as answer]
For received certificates, `--list-certs` only lists those in the in-memory cache, which has a default size of 32, see: https://github.com/strongswan/strongswan/blob/b47e9919f5c0beb6d809c0e65f531057487f758b/src/libstrongswan/credentials/sets/cert_cache.c#L27

**Reply by demahaj1** [reply]
Thanks @tobiasbrunner for your response !!!

Does this logic of default size of 32 applies only for **received certificates**?

I can see total 56 **local certificates** installed for 7 CA authority chains (Each chain having 8 Intermediate CA's), I have attached listed certificate logs..

Can you please confirm on this.. thanks

[swanctl_list_cert_output.txt](https://github.com/strongswan/strongswan/files/8089069/swanctl_list_cert_output.txt)

**Reply by tobiasbrunner** [reply]
Yes, you can have an unlimited amount of locally installed certificates. But note that responders send certificate requests for all trusted CA certificates (unless this is disabled completely), so having a lot of them installed might not be ideal as it could lead to IP fragmentation of IKE_SA_INIT (IKE fragmentation only works after IKE_SA_INIT).

Received certificates, which are not stored otherwise (well they are stored on the IKE_SAs themselves, unless _charon.flush_auth_cfg_ is enabled, but they are not provided in `--list-certs` from there) will just get removed from the cache if new certificate relationships are cached, which is why you might not see them anymore after a while.

Is this some kind of test? Or do you actually intend to use such deeply nested certificate chains in practice?

**Comment by demahaj1** [comment]
I am just testing this part, not using this kind of configuration with nested certificates..
Thanks a lot.. looks clear now regarding listing of certificates under swanctl .. 👍
